File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
    const dummyHead = new ListNode();
    dummyHead.next = head;
    
    let prev = dummyHead;
    let next = prev.next;
    let curr;
    if (!head) return head;
    curr = next.next ? next.next : null;
    
    while (next.next) {
        prev.next = curr;
        next.next = curr.next;
        curr.next = next;
        if (!next.next) return dummyHead.next;
        prev = next;
        next = next.next;
        curr = next.next ? next.next : null;
    }
    return dummyHead.next;
};

File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *   Val int
 *   Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
  if head == nil {
    return head
  }
  dh := &ListNode{}
  dh.Next = head
  prev := dh
  next := prev.Next
  curr := next.Next
  
  for next.Next != nil {
    prev.Next = curr
    next.Next = curr.Next
    curr.Next = next
    if next.Next == nil {
      break
    }
    prev = next
    next = next.Next
    curr = next.Next
  }
  return dh.Next
}